halo2/src/prover.rs

```diff
@@ -38,20 +42,38 @@ pub use halo2_proofs::poly::kzg::commitment::ParamsKZG;
 /// We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and Keccak256
 pub struct Halo2Prover {
     params: ParamsKZG<Bn256>,
+    vkey: Option<VerifyingKey<G1Affine>>,
 }
 
 impl Halo2Prover {
     pub fn new(size: DegreeType) -> Self {
         let degree = DegreeType::BITS - size.leading_zeros() + 1;
         Self {
             params: ParamsKZG::<Bn256>::new(degree),
+            vkey: None,
         }
     }
 
     pub fn new_from_setup(input: &mut impl io::Read) -> Result<Self, io::Error> {
         let params = ParamsKZG::<Bn256>::read(input)?;
 
-        Ok(Self { params })
+        Ok(Self { params, vkey: None })
+    }
+
+    pub fn add_verification_key<F: FieldElement>(
+        &mut self,
+        pil: &Analyzed<F>,
+        fixed: &[(String, Vec<F>)],
+        vkey: Vec<u8>,
+    ) {
+        let plaf = analyzed_to_plaf(pil, fixed);
+        let vkey: VerifyingKey<G1Affine> = VerifyingKey::<G1Affine>::from_bytes::<PlafH2Circuit>(
+            &vkey,
+            SerdeFormat::Processed,
+            plaf,
```

Comment:
@georgwiese `Default::default` actually doesn't work here, otherwise it complains that the circuit is empty (no constraints). So we need to pass the same circuit that we pass for the proof.
This causes this function to also require `pil` and `fixed`, because we need those to compute the `plaf` circuit.

Reply:
Oh OK. I think that makes sense. The circuit params I think are needed to be able to specify the constraints at runtime, and `Plaf` contains all information needed to do that,